File: scripts/hero-cross-browser.mjs
#!/usr/bin/env node
/**
 * Cross-browser validation for the hero headline spacing & clipping spans.
 *
 * Loads / in Chromium, Firefox, WebKit (desktop Safari), and WebKit with an
 * iPhone 14 viewport (mobile Safari). For each, asserts:
 *   - the h1 exists and contains exactly two .hero-reveal lines
 *   - neither line wraps (getClientRects().length === 1)
 *   - neither line overflows its container horizontally
 *   - computed letter/word spacing is non-zero (spacing tokens applied)
 *   - the clipping wrapper (.hero-reveal-clip / mask) is present and non-empty
 */
import { chromium, firefox, webkit, devices } from "playwright";
import { mkdirSync } from "node:fs";

const BASE_URL = process.env.BASE_URL ?? "http://localhost:8080";
const SHOT_DIR = "tests/visual/screenshots/cross-browser";
mkdirSync(SHOT_DIR, { recursive: true });

const TARGETS = [
  { name: "chromium-desktop",  type: chromium, contextOpts: { viewport: { width: 1280, height: 900 } } },
  { name: "firefox-desktop",   type: firefox,  contextOpts: { viewport: { width: 1280, height: 900 } } },
  { name: "webkit-desktop",    type: webkit,   contextOpts: { viewport: { width: 1280, height: 900 } } },
  { name: "webkit-iphone14",   type: webkit,   contextOpts: { ...devices["iPhone 14"] } },
  { name: "webkit-iphone-se",  type: webkit,   contextOpts: { ...devices["iPhone SE"] } },
];

async function inspect(page) {
  return page.evaluate(() => {
    const h1 = document.querySelector('[aria-label="Fitness Is The Only Law"]');
    if (!h1) return { error: "h1 not found" };
    const lines = Array.from(h1.querySelectorAll(".hero-reveal"));
    const vw = window.innerWidth;
    const out = { vw, lineCount: lines.length, lines: [] };
    for (const el of lines) {
      const rect = el.getBoundingClientRect();
      const cs = getComputedStyle(el);
      // Look for a clipping wrapper — prefer .hero-reveal-clip if present,
      // otherwise fall back to inspecting the reveal element itself.
      const clip = el.querySelector(".hero-reveal-clip") ?? el;
      const clipCS = getComputedStyle(clip);
      out.lines.push({
        text: el.textContent?.trim().slice(0, 60),
        width: Math.round(rect.width),
        height: Math.round(rect.height),
        clientRects: el.getClientRects().length,
        overflow: rect.right > vw + 0.5 || rect.left < -0.5,
        fontSize: cs.fontSize,
        letterSpacing: cs.letterSpacing,
        wordSpacing: cs.wordSpacing,
        clipOverflow: clipCS.overflow,
        clipHasContent: (clip.textContent ?? "").trim().length > 0,
      });
    }
    return out;
  });
}

function check(name, data) {
  const errs = [];
  if (data.error) errs.push(data.error);
  if (data.lineCount !== 2) errs.push(`expected 2 headline lines, got ${data.lineCount}`);
  data.lines?.forEach((l, i) => {
    if (l.clientRects !== 1) errs.push(`line ${i} wraps: ${l.clientRects} client rects`);
    if (l.overflow) errs.push(`line ${i} overflows viewport (w=${l.width}, vw=${data.vw})`);
    if (!l.clipHasContent) errs.push(`line ${i} clip wrapper has no text`);
    // letter/word spacing tokens should evaluate to non-zero px
    const ls = parseFloat(l.letterSpacing);
    const ws = parseFloat(l.wordSpacing);
    if (!Number.isFinite(ls)) errs.push(`line ${i} letterSpacing not numeric: ${l.letterSpacing}`);
    if (!Number.isFinite(ws)) errs.push(`line ${i} wordSpacing not numeric: ${l.wordSpacing}`);
  });
  return errs;
}

let failed = 0;
const summary = [];
for (const t of TARGETS) {
  let browser;
  try {
    browser = await t.type.launch({ headless: true });
    const context = await browser.newContext(t.contextOpts);
    const page = await context.newPage();
    await page.goto(BASE_URL, { waitUntil: "domcontentloaded" });
    await page.waitForSelector('[aria-label="Fitness Is The Only Law"]', { timeout: 10_000 });
    // Wait for fonts to settle so metrics reflect the final headline.
    await page.evaluate(() => document.fonts?.ready);
    await page.waitForTimeout(150);
    const data = await inspect(page);
    const errs = check(t.name, data);
    await page.locator('[aria-label="Fitness Is The Only Law"]').screenshot({
      path: `${SHOT_DIR}/${t.name}.png`,
    }).catch(() => {});
    const status = errs.length ? "FAIL" : "PASS";
    if (errs.length) failed++;
    summary.push({ target: t.name, status, vw: data.vw, lines: data.lines?.map(l => ({
      w: l.width, rects: l.clientRects, ls: l.letterSpacing, ws: l.wordSpacing,
    })), errors: errs });
    console.log(`[${status}] ${t.name}`, JSON.stringify(data.lines?.map(l => ({
      w: l.width, rects: l.clientRects, ls: l.letterSpacing, ws: l.wordSpacing, of: l.overflow,
    }))));
    if (errs.length) errs.forEach(e => console.log("   -", e));
  } catch (e) {
    failed++;
    summary.push({ target: t.name, status: "ERROR", error: String(e.message || e) });
    console.log(`[ERROR] ${t.name}:`, e.message);
  } finally {
    await browser?.close();
  }
}

console.log("\n=== SUMMARY ===");
for (const s of summary) console.log(s.status.padEnd(6), s.target, s.error ?? "");
process.exit(failed ? 1 : 0);

File: scripts/hero_cross_browser.py
"""Cross-browser hero headline validation (Chromium / Firefox / WebKit desktop & mobile)."""
import asyncio, json, os, sys
from pathlib import Path
from playwright.async_api import async_playwright

BASE_URL = os.environ.get("BASE_URL", "http://localhost:8080")
SHOT_DIR = Path("tests/visual/screenshots/cross-browser")
SHOT_DIR.mkdir(parents=True, exist_ok=True)

INSPECT_JS = """
() => {
  const h1 = document.querySelector('[aria-label="Fitness Is The Only Law"]');
  if (!h1) return { error: "h1 not found" };
  const lines = Array.from(h1.querySelectorAll('.hero-reveal'));
  const vw = window.innerWidth;
  return {
    vw,
    lineCount: lines.length,
    lines: lines.map(el => {
      const r = el.getBoundingClientRect();
      const cs = getComputedStyle(el);
      const clip = el.querySelector('.hero-reveal-clip') || el;
      return {
        text: (el.textContent || '').trim().slice(0, 60),
        width: Math.round(r.width),
        clientRects: el.getClientRects().length,
        overflow: r.right > vw + 0.5 || r.left < -0.5,
        letterSpacing: cs.letterSpacing,
        wordSpacing: cs.wordSpacing,
        fontSize: cs.fontSize,
        clipHasContent: (clip.textContent || '').trim().length > 0,
        clipOverflow: getComputedStyle(clip).overflow,
      };
    }),
  };
}
"""

def check(data):
    errs = []
    if data.get("error"): errs.append(data["error"]); return errs
    if data["lineCount"] != 2:
        errs.append(f"expected 2 lines, got {data['lineCount']}")
    for i, l in enumerate(data.get("lines", [])):
        if l["clientRects"] != 1:
            errs.append(f"line {i} wraps ({l['clientRects']} rects)")
        if l["overflow"]:
            errs.append(f"line {i} overflows viewport (w={l['width']}, vw={data['vw']})")
        if not l["clipHasContent"]:
            errs.append(f"line {i} clip wrapper has no text")
    return errs

async def run_target(p, name, launcher, context_opts):
    browser = await launcher.launch(headless=True)
    try:
        ctx = await browser.new_context(**context_opts)
        page = await ctx.new_page()
        await page.goto(BASE_URL, wait_until="domcontentloaded")
        await page.wait_for_selector('[aria-label="Fitness Is The Only Law"]', timeout=10_000)
        try: await page.evaluate("document.fonts && document.fonts.ready")
        except: pass
        await page.wait_for_timeout(200)
        data = await page.evaluate(INSPECT_JS)
        try:
            await page.locator('[aria-label="Fitness Is The Only Law"]').screenshot(
                path=str(SHOT_DIR / f"{name}.png"))
        except Exception: pass
        return data
    finally:
        await browser.close()

async def main():
    results = []
    async with async_playwright() as p:
        iphone14 = p.devices.get("iPhone 14") or p.devices["iPhone 13"]
        iphone_se = p.devices.get("iPhone SE") or p.devices["iPhone 8"]
        TARGETS = [
            ("chromium-desktop", p.chromium, {"viewport": {"width": 1280, "height": 900}}),
            ("firefox-desktop",  p.firefox,  {"viewport": {"width": 1280, "height": 900}}),
            ("webkit-desktop",   p.webkit,   {"viewport": {"width": 1280, "height": 900}}),
            ("webkit-iphone14",  p.webkit,   iphone14),
            ("webkit-iphone-se", p.webkit,   iphone_se),
        ]
        for name, launcher, opts in TARGETS:
            try:
                data = await run_target(p, name, launcher, opts)
                errs = check(data)
                status = "PASS" if not errs else "FAIL"
                summary = [{"w": l["width"], "rects": l["clientRects"],
                            "ls": l["letterSpacing"], "ws": l["wordSpacing"],
                            "of": l["overflow"]} for l in data.get("lines", [])]
                print(f"[{status}] {name} vw={data.get('vw')} {json.dumps(summary)}")
                for e in errs: print("   -", e)
                results.append((name, status, errs))
            except Exception as e:
                print(f"[ERROR] {name}: {e}")
                results.append((name, "ERROR", [str(e)]))
    print("\n=== SUMMARY ===")
    failed = 0
    for name, status, errs in results:
        print(f"{status:6} {name}")
        if status != "PASS": failed += 1
    sys.exit(1 if failed else 0)

asyncio.run(main())
